Add GDAKI test for nixlbench

aranadive opened the pull request:
- Introduced GDAKI options in configuration, including GPU device list and kernel parameters.
- Implemented GDAKI kernel functions for full and partial transfers in CUDA.
- Enhanced NIXL worker to manage GDAKI signal buffers and wireup notifications.
- Updated metadata exchange to support bidirectional communication in GDAKI mode.

This commit integrates GDAKI functionality into the NIXL worker. This a rebase of PR #785 to main.

aschwartz12 committed:
Add GDAKI test for nixlbench

- Introduced GDAKI options in configuration, including GPU device list and kernel parameters.
- Implemented GDAKI kernel functions for full and partial transfers in CUDA.
- Enhanced NIXL worker to manage GDAKI signal buffers and wireup notifications.
- Updated metadata exchange to support bidirectional communication in GDAKI mode.
- Update the calls in gdaki to use the new nixl GDAKI APIs. Also, update the signal used.
- Add a flag to check for nixl device api and fix up meson/src files.
- Update the readgpusignal call based on gpulevel correctly. Update other sites of gdaki api
  usage to be correct.
- Use the nixlGpuPostWriteXferReq API
- Use the getGpuSignalSize to get the right size for the signal.

This commit integrates GDAKI functionality into the NIXL worker

Signed-off-by: Asaf Schwartz <aschwartz@nvidia.com>
Signed-off-by: Adit Ranadive <aranadive@nvidia.com>

Signed-off-by: Adit Ranadive <aranadive@nvidia.com>

aranadive commented:
/build

aranadive commented:
/build

aranadive commented:
/build

aranadive commented:
/build

michal-shalev commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+    // Execute transfers for the specified number of iterations
+    for (int i = 0; i < num_iterations; i++) {
+        // Post the GPU transfer request with signal increment of 1
+        nixl_status_t status = nixlGpuPostSignalXferReq<nixl_gpu_level_t::BLOCK>(
```
The default in UCX is THREAD

michal-shalev replied:
Why do you use `nixlGpuPostSignalXferReq` and not `nixlGpuPostWriteXferReq`?

michal-shalev commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+                           const uint64_t signal_inc,
+                           const uint64_t remote_addr) {
+    constexpr size_t MAX_THREADS = 1024;
+    __shared__ nixlGpuXferStatusH xfer_status[MAX_THREADS];
```
AFAIK shared is needed here

aranadive replied:
Dont we need to get status from all threads if thread gpu_level is used?

aranadive replied:
Removed this

michal-shalev commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+    }
+}
+
+template<nixl_gpu_level_t level>
+__global__ void
+gdakiReadSignalKernel(const void *signal_addr, uint64_t *count) {
+    *count = nixlGpuReadSignal<level>(signal_addr);
+}
```
use `nixlGpuReadSignal`

michal-shalev commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+    // Check for launch errors
+    cudaError_t launch_error = cudaGetLastError();
+    if (launch_error != cudaSuccess) {
+        return NIXL_ERR_BACKEND;
```
add a log?

michal-shalev commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Why not the full copy right text?

```
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

aranadive replied:
CI now allows shorter 2-line copyright texts.

michal-shalev commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
same here:
```
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

aranadive replied:
We now allow shorter 2-line licenses in CI.

michal-shalev commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+ */
+
+#include <cuda_runtime.h>
+#include <nixl_types.h>
```
Are you sure you need to include `nixl_types.h`? `nixl_device.cuh` already includes it

michal-shalev reviewed (changes requested):
This implementation has several critical API usage issues and build configuration problems that need to be addressed:

1. Missing GPU signal APIs
This implementations missing the required `prepGpuSignal()` and `getGpuSignalSize()` APIs that are essential for GPU signal initialization.
2. Incorrect transfer API usage
The `gdakiFullTransferKernel` incorrectly uses `nixlGpuPostSignalXferReq()` instead of `nixlGpuPostWriteXferReq()` API. While the GPU transfer request already contains the IOV data from the original transfer request, (This is what you passed to `createGpuXferReq`) the signal-only API doesn't transfer the actual data, instead it only sends signals. 
3. Missing IOV parameters
Both `gdakiFullTransferKernel` and `gdakiPartialTransferKernel` are passing signal objects directly instead of providing the required IOV parameters (sizes, local addresses, remote addresses arrays) that the transfer APIs expect for data transfer operations.
4. Build configuration duplication
The benchmark's `meson.build` defines `-DHAVE_NIXL_DEV_API` when the root `meson.build` already provides `-DHAVE_UCX_GPU_DEVICE_API` and `ucx_gpu_device_api_available`.

michal-shalev commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+    }
+
+    // Launch partial transfer kernel based on coordination level
+    if (gpulevel == nixl_gpu_level_t::THREAD) {
+        gdakiPartialTransferKernel<nixl_gpu_level_t::THREAD>
+            <<<blocks_per_grid, threads_per_block, 0, stream>>>(
+                req_handle, num_iterations, signal_inc, remote_addr);
+    } else if (gpulevel == nixl_gpu_level_t::WARP) {
+        gdakiPartialTransferKernel<nixl_gpu_level_t::WARP>
+            <<<blocks_per_grid, threads_per_block, 0, stream>>>(
+                req_handle, num_iterations, signal_inc, remote_addr);
+    } else if (gpulevel == nixl_gpu_level_t::BLOCK) {
+        gdakiPartialTransferKernel<nixl_gpu_level_t::BLOCK>
+            <<<blocks_per_grid, threads_per_block, 0, stream>>>(
+                req_handle, num_iterations, signal_inc, remote_addr);
+    }
```
Is this redundant considering that you check `gpu_level` when you call `launchGdakiPartialKernel` from `nixl_worker.cpp`?

michal-shalev replied:
Also here you allow block level but in `nixl_worker.cpp` when block level is requested, you use `launchGdakiKernel` instead, why?

aranadive replied:
Not redundant since there cant use the runtime level directly for the templated level. The block level is dead code since not passed. i replaced that with an error.

aranadive commented on benchmark/nixlbench/meson.build:
```diff
@@ -148,6 +155,33 @@ if cuda_fabric_available
+        link_args : ['-L' + ucx_lib_path, '-lucp', '-lucs', '-luct'],
+        compile_args : ['-I' + ucx_inc_path])
+    if cuda_available and add_languages('cuda', required: false)
+	nvcc = meson.get_compiler('cuda')
```
Done

aranadive commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+    } else if (level == nixl_gpu_level_t::WARP) {
+        gdakiReadSignalKernel<nixl_gpu_level_t::WARP><<<1, 32>>>(signal_addr, d_count);
+    } else if (level == nixl_gpu_level_t::BLOCK) {
+        gdakiReadSignalKernel<nixl_gpu_level_t::BLOCK><<<1, 1>>>(signal_addr, d_count);
```
Not sure, if we are reading a single value here.

aranadive commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cuh:
```diff
@@ -0,0 +1,45 @@
+#include <nixl_types.h>
+
+#ifdef __cplusplus
+extern "C" {
```
Probably cannot remove extern since this is built as a separate cuda kernel object and then linked.

aranadive commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+    if (gpulevel == nixl_gpu_level_t::BLOCK) {
+        gdakiFullTransferKernel<<<blocks_per_grid, threads_per_block, 0, stream>>>(
+            req_handle, num_iterations, signal_inc, remote_addr);
+    } else {
```
I added a log message here and removed the if/else.

aranadive commented:
> This implementation has several critical API usage issues and build configuration problems that need to be addressed:
> 
> 1. Missing GPU signal APIs
>    This implementations missing the required `prepGpuSignal()` and `getGpuSignalSize()` APIs that are essential for GPU signal initialization.
> 2. Incorrect transfer API usage
>    The `gdakiFullTransferKernel` incorrectly uses `nixlGpuPostSignalXferReq()` instead of `nixlGpuPostWriteXferReq()` API. While the GPU transfer request already contains the IOV data from the original transfer request, (This is what you passed to `createGpuXferReq`) the signal-only API doesn't transfer the actual data, instead it only sends signals.
> 3. Missing IOV parameters
>    Both `gdakiFullTransferKernel` and `gdakiPartialTransferKernel` are passing signal objects directly instead of providing the required IOV parameters (sizes, local addresses, remote addresses arrays) that the transfer APIs expect for data transfer operations.
> 4. Build configuration duplication
>    The benchmark's `meson.build` defines `-DHAVE_NIXL_DEV_API` when the root `meson.build` already provides `-DHAVE_UCX_GPU_DEVICE_API` and `ucx_gpu_device_api_available`.

1, 2, 3 should be addressed now. For 4, we need a different macro anyway (we need to check if NIXL supports device API) and also since nixlbench is a separate meson project under nixl git.

michal-shalev commented on benchmark/nixlbench/src/utils/utils.cpp:
```diff
@@ -184,6 +212,52 @@ xferBenchConfig::loadFromFlags() {
+                }
+            }
+
+            if (gdaki_threads_per_block < 1 || gdaki_threads_per_block > 1024) {
```
please add constants instead of magic numbers

aranadive replied:
Done

tvegas1 commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,228 @@
+        } while (status == NIXL_IN_PROG);
+
+        if (status != NIXL_SUCCESS) {
+            return; // Early exit on error
```
shall we add printf as this should not happen in normal run?

brminich commented on benchmark/nixlbench/src/worker/nixl/nixl_worker.cpp:
```diff
@@ -1024,31 +1433,78 @@ xferBenchNixlWorker::transfer(size_t block_size,
+        // Wait for warmup iterations to complete
+        uint64_t count = 0;
+        while (count < skip) {
+            count = readNixlGpuSignal(signal_addr, gpu_level.data());
```
can we do all polling logic in the kernel? Or need to avoid cudaMalloc/free for every read

brminich commented on benchmark/nixlbench/src/worker/nixl/nixl_worker.cpp:
```diff
@@ -893,20 +1112,203 @@ xferBenchNixlWorker::exchangeIOV(const std::vector<std::vector<xferBenchIOV>> &l
+        unsigned int idx = 0;
+
+        desc_cnt = local_iov.size();
+        local_addrs = (void **)malloc(desc_cnt * sizeof(void *));
+        remote_addrs = (uint64_t *)malloc(desc_cnt * sizeof(uint64_t));
+        lengths = (size_t *)malloc(desc_cnt * sizeof(size_t));
+
```
can we use vectors instead?

brminich commented on benchmark/nixlbench/src/worker/nixl/nixl_worker.cpp:
```diff
@@ -893,20 +1112,203 @@ xferBenchNixlWorker::exchangeIOV(const std::vector<std::vector<xferBenchIOV>> &l
+
+        uint64_t remote_addr = 0;
+
+        params.notifMsg = "0xBEEF";
+        params.hasNotif = true;
+        target = "target";
```
do we really have notif?

brminich commented on benchmark/nixlbench/src/worker/nixl/nixl_worker.cpp:
```diff
@@ -561,6 +572,183 @@ xferBenchNixlWorker::cleanupBasicDescObj(xferBenchIOV &iov) {
+};
+
+int
+xferBenchNixlWorker::sendWireupMessage() {
```
why is it needed?

brminich commented on benchmark/nixlbench/src/worker/nixl/nixl_worker.cpp:
```diff
@@ -687,10 +875,37 @@ xferBenchNixlWorker::allocateMemory(int num_threads) {
+        // Add signal buffer for GDAKI if enabled
+        if (is_gdaki_enabled) {
+#ifdef HAVE_NIXL_DEV_API
+            size_t sigsize;
+            nixl_opt_args_t extra_params = {.backends = {backend_engine}};
+
+            CHECK_NIXL_ERROR(agent->getGpuSignalSize(sigsize, &extra_params),
+                             "getGpuSignalSize failed");
+
+            std::optional<xferBenchIOV> signal_desc = initSignalBuffer(0, sigsize);
+            if (signal_desc) {
+                iov_list.push_back(signal_desc.value());
+                signal_buffers.push_back(signal_desc.value());
```
can it be done in some specific device api method instead of general allocation one?

iyastreb commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cuh:
```diff
@@ -0,0 +1,56 @@
+                          void *const *local_addrs,
+                          const uint64_t *remote_addrs,
+                          int threads_per_block = 256,
+                          int blocks_per_grid = 1,
```
Minor: Some args are passed by const, some not.. I propose to either make all of them const, or remove const from all of them

iyastreb commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cuh:
```diff
@@ -0,0 +1,56 @@
+// Launch Device kernel for partial transfers (supports thread/warp/block coordination)
+nixl_status_t
+launchDevicePartialKernel(nixlGpuXferReqH *req_handle,
+                          int num_iterations,
```
11 args is too much, why not to join them in the structure something like *_params?
This way you can reuse it for both kernels

iyastreb commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,418 @@
+
+// Helper function to get request index based on coordination level (from gtest)
+template<nixl_gpu_level_t level>
+__device__ constexpr size_t
```
inline / forceinline

iyastreb commented on benchmark/nixlbench/src/worker/nixl/nixl_worker.cpp:
```diff
@@ -561,6 +572,183 @@ xferBenchNixlWorker::cleanupBasicDescObj(xferBenchIOV &iov) {
+
+#define NOTIF_WIREUP_COMP "wireup_complete"
+
+int
```
Why it does not return nixl_status?

iyastreb commented on benchmark/nixlbench/src/worker/nixl/nixl_worker.cpp:
```diff
@@ -894,20 +1114,204 @@ xferBenchNixlWorker::exchangeIOV(const std::vector<std::vector<xferBenchIOV>> &l
+        unsigned int idx = 0;
+
+        desc_cnt = local_iov.size();
+        local_addrs = (void **)malloc(desc_cnt * sizeof(void *));
```
malloc is bad choice, very verbose, not expcetion safe, copy paste...
Just use std::vector

aranadive commented on benchmark/nixlbench/src/utils/utils.cpp:
```diff
@@ -186,6 +214,53 @@ xferBenchConfig::loadFromFlags() {
+            if (gdaki_threads_per_block < 1 ||
+                gdaki_threads_per_block > XFERBENCH_DEV_API_MAX_THREADS) {
+                std::cerr << "Invalid GDAKI threads per block: " << gdaki_threads_per_block
+                          << ". Must be between 1 and 1024" << std::endl;
```
Done

aranadive commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,264 @@
+                                                                     signal,
+                                                                     1,
+                                                                     true,
+                                                                     &xfer_status);
```
What is different? The parameters specified? Hmm most of that alignment was aligned as per clang.

aranadive commented on benchmark/nixlbench/src/worker/nixl/gdaki_kernels.cu:
```diff
@@ -0,0 +1,418 @@
+                        int blocks_per_grid) {
+    // Validate parameters
+    if (num_iterations <= 0 || req_handle == nullptr) {
+        return NIXL_ERR_INVALID_PARAM;
```
Done

aranadive committed:
Addressed comments

Signed-off-by: Adit Ranadive <aranadive@nvidia.com>

aranadive committed:
More fixes

Signed-off-by: Adit Ranadive <aranadive@nvidia.com>

aranadive commented on benchmark/nixlbench/src/worker/nixl/nixl_worker.h:
```diff
@@ -57,8 +62,18 @@ class xferBenchNixlWorker: public xferBenchWorker {
         void
         poll(size_t block_size) override;
+        void
+        device_poll(size_t block_size, unsigned int skip, unsigned int num_iter);
```
Done

aranadive committed:
More comments addressed

Signed-off-by: Adit Ranadive <aranadive@nvidia.com>

aranadive committed:
Fix clang

Signed-off-by: Adit Ranadive <aranadive@nvidia.com>

aranadive commented:
/build

aranadive committed:
Fix copyright

Signed-off-by: Adit Ranadive <aranadive@nvidia.com>

aranadive committed:
Fix build issues

Added ucx include path explicitly for nixlbench to find headers for
device initiated comms.

Also, copied the GPU coordination levels for device code to new file to
avoid including utils header when building device code.

Signed-off-by: Adit Ranadive <aranadive@nvidia.com>
